src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Migrations/Custom/EES1598RemoveChartLabelsFieldTests.cs

```diff
@@ -0,0 +1,599 @@
+                                    LineStyle = ChartLineStyle.solid
+                                }
+                            }
+                        },
+                        Axes = new Dictionary<string, ChartAxisConfiguration>
+                        {
+                            {
+                                "major",
+                                new ChartAxisConfiguration
+                                {
+                                    DataSets = new List<ChartDataSet>
+                                    {
+                                        new ChartDataSet
+                                        {
+                                            Filters = new List<Guid> { filter1Id },
+                                            Indicator = indicator1Id
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                },
+            };
+
+            var contextId = Guid.NewGuid().ToString();
+
+            using (var context = ContentDbUtils.InMemoryContentDbContext(contextId))
+            {
+                context.Add(dataBlock);
+                context.SaveChanges();
+            }
+
+            using (var context = ContentDbUtils.InMemoryContentDbContext(contextId))
+            {
+                var migration = new EES1598RemoveChartLabelsField(context, isTest: true);
+                migration.Apply();
+            }
+
+            using (var context = ContentDbUtils.InMemoryContentDbContext(contextId))
+            {
+                var savedDataBlock = context.DataBlocks.Find(dataBlock.Id);
+
+                Assert.Null(savedDataBlock.Charts[0].Labels);
+
+                var dataSets = savedDataBlock.Charts[0].Axes?["major"].DataSets;
+
+                Assert.NotNull(dataSets);
+                Assert.Equal(2, dataSets.Count);
+
+                Assert.Equal(indicator1Id,  dataSets[0].Indicator);
+                Assert.Equal(new List<Guid>{ filter1Id },  dataSets[0].Filters);
+
+                Assert.Equal("Test colour 1", dataSets[0].Config.Colour);
+                Assert.Equal("Test label 1", dataSets[0].Config.Label);
+                Assert.Equal(ChartLineSymbol.diamond, dataSets[0].Config.Symbol);
+                Assert.Equal(ChartLineStyle.dotted, dataSets[0].Config.LineStyle);
+
+                Assert.Equal(indicator2Id,  dataSets[1].Indicator);
+                Assert.Equal(new List<Guid> { filter2Id },  dataSets[1].Filters);
+
+                Assert.Equal("Test colour 2", dataSets[1].Config.Colour);
+                Assert.Equal("Test label 2", dataSets[1].Config.Label);
+                Assert.Equal(ChartLineSymbol.square, dataSets[1].Config.Symbol);
+                Assert.Equal(ChartLineStyle.solid, dataSets[1].Config.LineStyle);
+            }
+        }
+
+        [Fact]
+        public void IgnoresJunkLabels()
```

Comment:
What are "junk labels"?  Are they just leftover data from a bygone era?

Reply:
Yep, major technical debt. Don't want to point too many fingers but the original chart implementor decided to try and match data sets to labels using strings with all the ids concatenated together with underscores e.g. `some-indicator-id_some-filter-id-1_some-filter-id-2_some-location-1_some-location-2_some-time-period`. Even more confusingly, in some instances, it's just the time period by itself - hence why some labels are called 'junk'.

After I took over the charts code, I refactored the implementation to remove these horrendous strings and serialize it properly as a data set. Fortunately it was a bit half baked, so I was able to put in some code that allowed us to continue rendering charts with this older format, whilst new charts would just use the new format.

The upcoming legends re-labelling feature may require some additional matching on data sets, so I figured it was the right time to finally get the ball moving on removing the old labels implementation.

Reply:
Cool cool thanks for all the info.